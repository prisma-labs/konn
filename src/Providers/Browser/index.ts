export * as Browser from './Browser'
