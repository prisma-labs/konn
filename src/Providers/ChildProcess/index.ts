export * as ChildProcess from './ChildProcess'
