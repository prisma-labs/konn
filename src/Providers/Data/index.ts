export * as Data from './Data'
