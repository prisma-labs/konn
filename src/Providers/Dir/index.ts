export * as Dir from './Dir'
