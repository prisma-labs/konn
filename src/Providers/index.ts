export * as Providers from './index_'
