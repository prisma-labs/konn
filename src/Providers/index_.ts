export * from './Dir'
export * from './Run'
