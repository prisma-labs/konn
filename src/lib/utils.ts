/**
 * Does the todo thing.
 *
 * @example
 *
 *   const a = foobar({
 *     a: 1,
 *     b: 2,
 *   })
 *
 */
export function todo(): void {
  console.log(nameof(todo))
}
