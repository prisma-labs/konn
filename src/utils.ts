export type MaybePromise<T> = void | Promise<T>
