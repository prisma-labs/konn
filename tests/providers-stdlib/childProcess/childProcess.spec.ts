import { casesHandled } from 'floggy/dist-cjs/utils'
import * as Fs from 'fs-jetpack'
import { konn, providers } from '~/index'
import { Providers } from '~/Providers'
import { timeout } from '~/utils'
import { tests } from './__data__'

jest.setTimeout(30_000)

it('exports shorthand access', () => {
  expect(typeof providers.childProcess).toEqual('function')
})

it('exports longhand access', () => {
  expect(typeof Providers.ChildProcess.create).toEqual('function')
})

/**
 * Test that expected types are exported
 */
type _Types = [
  Providers.ChildProcess.Needs,
  Providers.ChildProcess.Contributes,
  Providers.ChildProcess.Params
]

describe('runs while fixture process is running', () => {
  const ctx = konn()
    .useBeforeAll(
      providers.childProcess({
        command: `ts-node ./tests/providers-stdlib/childProcess/__long_running.fixture.ts`,
        start: /ready/,
      })
    )
    .done()
  it('test', () => {
    const filePath = './tests/fixture-process-proof-file.txt'
    expect(Fs.exists(filePath)).toBeTruthy()
    Fs.remove(filePath)
  })
})

describe('long running node script without signal handling works', () => {
  const ctx = konn()
    .useBeforeAll(
      providers.childProcess({
        command: `ts-node ./tests/providers-stdlib/childProcess/__long_running_no_signal_handling.fixture.ts`,
        start: /ready/,
      })
    )
    .done()
  it('test', async () => {
    await timeout(100)
  })
})

describe('error', () => {
  Object.values(tests).map((_) => {
    it(_.replace(/_/g, ' '), () => {
      const stdio = Fs.read(`tests/providers-stdlib/childProcess/__${_}.log.ansi`, 'utf8')!
      expect(stdio).toMatch(/Test Suites: \d failed, \d total/)
      switch (_) {
        case 'error_on_exit':
          expect(stdio).toMatch(/Error: Something bad happened while reacting to sigterm/)
          break
        case 'error_on_spawn':
          expect(stdio).toMatch(/Error while attempting to spawn: spawn this-will-fail-on-spawn ENOENT/)
          break
        case 'start_timeout':
          expect(stdio).toMatch(
            /Timed out \(500 ms\) while waiting for child process start signal \/ready\/. While waiting, saw this stdout and stderr output from the process:/
          )
          // TODO
          // In CI this does not work on macOS for some reason...
          // There, it shows the message: N/A -- THERE WAS NO STDOUT/STDERR FROM THE CHILD PROCESS!
          // expect(stdio).toMatch(/-+\n *1 \(stdout\)\n *2 \(stdout\)\n *3 \(stderr\)\n *-+/)
          break
        case 'error_while_running':
          expect(stdio).toMatch(/Error: Something went wrong while running./)
          break
        default:
          casesHandled(_)
      }
    })
  })
})
