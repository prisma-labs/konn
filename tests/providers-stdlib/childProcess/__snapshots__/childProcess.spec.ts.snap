// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error error on exit 1`] = `
"FAIL tests/providers-stdlib/childProcess/__error_on_exit.ts


  ● Test suite failed to run

    Test context \\"childProcess\\" failed while running setdown logic inside Jest \\"afterAll\\": 
    The child process exited with non-zero exit code: 1. Its combined stderr and stdout output was:

    ----------------------------------------------------------------
    ready
    /DYNAMIC FILE PATH
        throw new Error('Something bad happened while reacting to sigterm.');
              ^
    Error: Something bad happened while reacting to sigterm.

      12 | const workingTimeout = setTimeout(() => {}, 1000 * 60 * 60 /* 1h */)
      13 |
    > 14 | process.on('SIGTERM', () => {
         |           ^
      15 |   clearTimeout(workingTimeout)
      16 |   throw new Error('Something bad happened while reacting to sigterm.')
      17 | })

      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE
      ---------------------------------------------------------------- 
      Command failed with exit code 1: ts-node ./tests/providers-stdlib/childProcess/__error_on_exit.fixture.ts
      /DYNAMIC FILE PATH
          throw new Error('Something bad happened while reacting to sigterm.');
                ^
      Error: Something bad happened while reacting to sigterm.
      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE
      Error: The child process exited with non-zero exit code: 1. Its combined stderr and stdout output was:
      ----------------------------------------------------------------
      ready
      /DYNAMIC FILE PATH
          throw new Error('Something bad happened while reacting to sigterm.');
                ^
      Error: Something bad happened while reacting to sigterm.
      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE
      ---------------------------------------------------------------- 
      Command failed with exit code 1: ts-node ./tests/providers-stdlib/childProcess/__error_on_exit.fixture.ts
      /DYNAMIC FILE PATH
          throw new Error('Something bad happened while reacting to sigterm.');
                ^
      Error: Something bad happened while reacting to sigterm.
      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE
      Error: Command failed with exit code 1: ts-node ./tests/providers-stdlib/childProcess/__error_on_exit.fixture.ts
      /DYNAMIC FILE PATH
          throw new Error('Something bad happened while reacting to sigterm.');
                ^
      Error: Something bad happened while reacting to sigterm.
      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE

Test Suites: 1 failed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time: DYNAMIC NUMBER
Ran all test suites.
"
`;

exports[`error error while running 1`] = `
"FAIL tests/providers-stdlib/childProcess/__error_while_running.ts


  ● Test suite failed to run

    Test context \\"childProcess\\" failed while running setdown logic inside Jest \\"afterAll\\": 
    The child process exited with non-zero exit code: 1. Its combined stderr and stdout output was:

    ----------------------------------------------------------------
    ready
    /DYNAMIC FILE PATH
        throw new Error('Something went wrong while running.');
              ^
    Error: Something went wrong while running.

      2 |
      3 | setTimeout(() => {
    > 4 |   throw new Error('Something went wrong while running.')
        |           ^
      5 | }, 100)
      6 |
      7 | process.on('SIGTERM', () => {

      at DYNAMIC STACK TRACE LINE
      ---------------------------------------------------------------- 
      Command failed with exit code 1: ts-node ./tests/providers-stdlib/childProcess/__error_while_running.fixture.ts
      at DYNAMIC STACK TRACE LINE
      Error: The child process exited with non-zero exit code: 1. Its combined stderr and stdout output was:
      ----------------------------------------------------------------
      ready
      /DYNAMIC FILE PATH
          throw new Error('Something went wrong while running.');
                ^
      Error: Something went wrong while running.
      at DYNAMIC STACK TRACE LINE
      ---------------------------------------------------------------- 
      Command failed with exit code 1: ts-node ./tests/providers-stdlib/childProcess/__error_while_running.fixture.ts
      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE
      Error: Command failed with exit code 1: ts-node ./tests/providers-stdlib/childProcess/__error_while_running.fixture.ts
      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE

Test Suites: 1 failed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time: DYNAMIC NUMBER
Ran all test suites.
"
`;

exports[`error start timeout 1`] = `
"FAIL tests/providers-stdlib/childProcess/__start_timeout.ts
  start timeout
    ✕ test

  ● start timeout › test

    Test context \\"childProcess\\" failed while running setup logic inside Jest \\"beforeAll\\": 
    Timed out (500 ms) while waiting for child process start signal /ready/. While waiting, saw this stdout and stderr output from the process:

    ----------------------------------------------------------------
    1 (stdout)
    2 (stdout)
    3 (stderr)
    ----------------------------------------------------------------

      30 |       })
      31 |     } catch (error) {
    > 32 |       throw ono(
         |                ^
      33 |         error as Error,
      34 |         \`Test context \\"\${params.providerName}\\" failed while running setup logic inside Jest \\"\${params.jestHookName}\\":\`
      35 |       )

      at DYNAMIC STACK TRACE LINE
      Error: Timed out (500 ms) while waiting for child process start signal /ready/. While waiting, saw this stdout and stderr output from the process:
      ----------------------------------------------------------------
      1 (stdout)
      2 (stdout)
      3 (stderr)
      ----------------------------------------------------------------
      at DYNAMIC STACK TRACE LINE
      at DYNAMIC STACK TRACE LINE

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
Snapshots:   0 total
Time: DYNAMIC NUMBER
Ran all test suites.
"
`;
